common/lib/dependabot/command_helpers.rb

```diff
@@ -0,0 +1,202 @@
+      end
+
+      # Return the PID of the process (if available)
+      sig { returns(T.nilable(Integer)) }
+      def pid
+        @process_status.pid
+      end
+
+      sig { returns(T.nilable(Integer)) }
+      def termsig
+        @process_status.termsig
+      end
+
+      # String representation of the status
+      sig { returns(String) }
+      def to_s
+        if @custom_exitstatus
+          "pid #{pid || 'unknown'}: exit #{@custom_exitstatus} (custom status)"
+        else
+          @process_status.to_s
+        end
+      end
+    end
+
+    DEFAULT_TIMEOUTS = T.let({
+      no_time_out: -1,  # No timeout
+      local: 30,        # Local commands
+      network: 120,     # Network-dependent commands
+      long_running: 300 # Long-running tasks (e.g., builds)
+    }.freeze, T::Hash[T.untyped, T.untyped])
+
+    # rubocop:disable Metrics/AbcSize
+    # rubocop:disable Metrics/MethodLength
+    # rubocop:disable Metrics/PerceivedComplexity
+    # rubocop:disable Metrics/CyclomaticComplexity
+    sig do
+      params(
+        env_cmd: T::Array[T.any(T::Hash[String, String], String)],
+        stdin_data: T.nilable(String),
+        stderr_to_stdout: T::Boolean,
+        command_type: Symbol,
+        timeout: Integer
+      ).returns([T.nilable(String), T.nilable(String), T.nilable(ProcessStatus), Float])
+    end
+    def self.capture3_with_timeout(
+      env_cmd,
+      stdin_data: nil,
+      stderr_to_stdout: false,
+      command_type: :network,
+      timeout: -1
+    )
+      # Assign default timeout based on command type if timeout < 0
+      timeout = DEFAULT_TIMEOUTS[command_type] if timeout.negative?
+
+      stdout = T.let("", String)
+      stderr = T.let("", String)
+      status = T.let(nil, T.nilable(ProcessStatus))
+      pid = T.let(nil, T.untyped)
+      start_time = Time.now
+
+      begin
+        T.unsafe(Open3).popen3(*env_cmd) do |stdin, stdout_io, stderr_io, wait_thr|
+          pid = wait_thr.pid
+          stdin&.write(stdin_data) if stdin_data
+          stdin&.close
+
+          stdout_io.sync = true
+          stderr_io.sync = true
+
+          # Array to monitor both stdout and stderr
+          ios = [stdout_io, stderr_io]
+
+          last_output_time = Time.now # Track the last time output was received
+
+          until ios.empty?
+            # Calculate remaining timeout dynamically
+            remaining_timeout = timeout - (Time.now - last_output_time)
+
+            # Raise an error if timeout is exceeded
+            if remaining_timeout <= 0
+              terminate_process(pid)
+              status = ProcessStatus.new(wait_thr.value, 124)
+              raise Timeout::Error, "Timed out due to inactivity after #{timeout} seconds"
+            end
+
+            # Use IO.select with a dynamically calculated short timeout
+            ready_ios = IO.select(ios, nil, nil, [0.1, remaining_timeout].min)
```

Comment:
We are not monitoring for write readiness (`nil`) because writing is already completed at the beginning, and we do not perform incremental writes. Similarly, we do not monitor error conditions (`nil`) as they are not required in this context. For read readiness, we monitor the output streams (`stdout` and `stderr`) using a timeout to avoid blocking indefinitely. The timeout is set to the shorter of 0.1 seconds or the remaining timeout (`[0.1, remaining_timeout].min`), ensuring responsiveness while respecting the overall timeout limit.